rate.go

```diff
@@ -1,116 +1,237 @@
 	}
+
+	return &GCRARateLimiter{
+		delayVariationTolerance: quota.MaxRate.period * (time.Duration(quota.MaxBurst) + 1),
+		emissionInterval:        quota.MaxRate.period,
+		limit:                   quota.MaxBurst + 1,
+		store:                   st,
+	}, nil
 }
 
-// Limit is called for each request to the throttled handler. It checks if
-// the request can go through and signals it via the returned channel.
-// It returns an error if the operation fails.
-func (r *rateLimiter) Limit(w http.ResponseWriter, req *http.Request) (<-chan bool, error) {
-	// Create return channel and initialize
-	ch := make(chan bool, 1)
-	ok := true
-	key := r.vary.Key(req)
-
-	// Get the current count and remaining seconds
-	cnt, secs, err := r.store.Incr(key, r.window)
-	// Handle the possible situations: error, begin new window, or increment current window.
-	switch {
-	case err != nil && err != ErrNoSuchKey:
-		// An unexpected error occurred
-		return nil, err
-	case err == ErrNoSuchKey || secs <= 0:
-		// Reset counter
-		if err := r.store.Reset(key, r.window); err != nil {
-			return nil, err
+// RateLimit checks whether a particular key has exceeded a rate
+// limit. It also returns a RateLimitContext to provide additional
+// information about the state of the RateLimiter.
+//
+// If the rate limit has not been exceeded, the underlying storage is
+// updated by the supplied quantity. For example, a quantity of 1
+// might be used to rate limit a single request while a greater
+// quantity could rate limit based on the size of a file upload in
+// megabytes. If quantity is 0, no update is performed allowing you
+// to "peek" at the state of the RateLimiter for a given key.
+func (g *GCRARateLimiter) RateLimit(key string, quantity int) (bool, RateLimitContext, error) {
+	var tat, newTat, now time.Time
+	var ttl time.Duration
+	rlc := RateLimitContext{Limit: g.limit, RetryAfter: -1}
+	limited := false
+
+	i := 0
+	for {
+		var err error
+		var tatVal int64
+		var updated bool
+
+		// tat refers to the theoretical arrival time that would be expected
+		// from equally spaced requests at exactly the rate limit.
+		tatVal, now, err = g.store.GetWithTime(key)
+		if err != nil {
+			return false, rlc, err
+		}
+
+		if tatVal == -1 {
+			tat = now
+		} else {
+			tat = time.Unix(0, tatVal)
+		}
+
+		increment := time.Duration(quantity) * g.emissionInterval
+		if now.After(tat) {
+			newTat = now.Add(increment)
+		} else {
+			newTat = tat.Add(increment)
+		}
+
+		// Block the request if the next permitted time is in the future
+		allowAt := newTat.Add(-(g.delayVariationTolerance))
+		if diff := now.Sub(allowAt); diff < 0 {
+			if increment <= g.delayVariationTolerance {
+				rlc.RetryAfter = -diff
+			}
+			ttl = tat.Sub(now)
+			limited = true
+			break
+		}
+
+		ttl = newTat.Sub(now)
+
+		if tatVal == -1 {
+			updated, err = g.store.SetIfNotExistsWithTTL(key, newTat.UnixNano(), ttl)
```

Comment:
It looks like you [accounted for some of my original feedback on this one](https://github.com/throttled/throttled/pull/8#discussion_r35445148). +1. I have a hard time imagining how a race condition would occur here at this point.